core/src/layers/prometheus_client.rs

```diff
@@ -0,0 +1,669 @@
+///
+///     // Export prometheus metrics.
+///     let mut buf = String::new();
+///     prometheus_client::encoding::text::encode(&mut buf, &registry).unwrap();
+///     println!("## Prometheus Metrics");
+///     println!("{}", buf);
+///     Ok(())
+/// }
+/// ```
+#[derive(Debug)]
+pub struct PrometheusClientLayer {
+    metrics: PrometheusClientMetrics,
+}
+
+impl PrometheusClientLayer {
+    /// create PrometheusClientLayer while registering itself to this registry.
+    pub fn with_registry(registry: &mut Registry) -> Self {
+        let metrics = PrometheusClientMetrics::register(registry);
+        Self { metrics }
+    }
+}
+
+impl<A: Accessor> Layer<A> for PrometheusClientLayer {
+    type LayeredAccessor = PrometheusAccessor<A>;
+
+    fn layer(&self, inner: A) -> Self::LayeredAccessor {
+        let meta = inner.info();
+        let scheme = meta.scheme();
+
+        let stats = Arc::new(self.metrics.clone());
+        PrometheusAccessor {
+            inner,
+            stats,
+            scheme: scheme.to_string(),
+        }
+    }
+}
+
+type VecLabels = Vec<(&'static str, String)>;
+
+/// [`PrometheusClientMetrics`] provide the performance and IO metrics with the `prometheus-client` crate.
+#[derive(Debug, Clone)]
+struct PrometheusClientMetrics {
+    /// Total counter of the specific operation be called.
+    requests_total: Family<VecLabels, Counter>,
+    /// Latency of the specific operation be called.
+    request_duration_seconds: Family<VecLabels, Histogram>,
+    /// The histogram of bytes
+    bytes_histogram: Family<VecLabels, Histogram>,
+}
+
+impl PrometheusClientMetrics {
+    pub fn register(registry: &mut Registry) -> Self {
+        let requests_total = Family::default();
+        let request_duration_seconds = Family::<VecLabels, _>::new_with_constructor(|| {
+            let buckets = prometheus_client::metrics::histogram::exponential_buckets(0.01, 2.0, 16);
+            Histogram::new(buckets)
+        });
+        let bytes_histogram = Family::<VecLabels, _>::new_with_constructor(|| {
+            let buckets = prometheus_client::metrics::histogram::exponential_buckets(1.0, 2.0, 16);
+            Histogram::new(buckets)
+        });
+
+        registry.register("requests_total", "", requests_total.clone());
+        registry.register(
+            "request_duration_seconds",
+            "",
+            request_duration_seconds.clone(),
+        );
+        registry.register("bytes_histogram", "", bytes_histogram.clone());
+        Self {
+            requests_total,
+            request_duration_seconds,
+            bytes_histogram,
+        }
+    }
+
+    fn increment_errors_total(&self, op: Operation, kind: ErrorKind) {
+        let labels = vec![("operation", op.to_string()), ("kind", kind.to_string())];
+        self.requests_total.get_or_create(&labels).inc();
+    }
+
+    fn increment_request_total(&self, scheme: &str, op: Operation) {
+        let labels = vec![
```

Comment:
EncodeLabelValue is quite invasive, we have to implement `EncodeLabelValue` for `Scheme` and `Operation` in their crate.

I made another way that use `&[(&'static str, &'static str); 2]` as label set to avoid the copying, PTAL